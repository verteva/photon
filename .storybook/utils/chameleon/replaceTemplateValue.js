import parsers from './parsers';
import { resolveTemplateString } from './resolveTemplateString';

export const replaceTemplateValue = (templateValue, root) => {
  if (typeof templateValue === 'string') {
    return resolveTemplateString(templateValue, root);
  }
  if (typeof templateValue === 'object') {
    if (templateValue.type) {
      console.log('TYPE', templateValue.type);
      console.log('VALUE FOR OTHER', templateValue);
      const resolver = parsers[templateValue.type];
      return resolver(templateValue, root);
    }
  }
};
