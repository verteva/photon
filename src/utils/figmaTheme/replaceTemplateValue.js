import parsers from './parsers';
import { resolveTemplateString } from './resolveTemplateString';

export const replaceTemplateValue = (templateValue, root) => {
  if (typeof templateValue === 'string') {
    return resolveTemplateString(templateValue, root);
  }
  if (typeof templateValue === 'object') {
    if (templateValue.type) {
      const resolver = parsers[templateValue.type];
      return resolver(templateValue, root);
    }
  }
};
