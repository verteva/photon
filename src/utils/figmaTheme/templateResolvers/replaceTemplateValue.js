import parsers from './parsers';
import { resolveTemplateString } from './resolveTemplateString';

export const replaceTemplateValue = (templateValue, root) => {
  if (typeof templateValue !== 'string' && typeof templateValue !== 'object') {
    throw new Error(
      'Invalid template value, must be a string or an object with a type property'
    );
  }
  if (typeof templateValue === 'object' && !templateValue.type) {
    throw new Error(
      'Invalid template value, cant parse objects without a type'
    );
  }
  if (typeof templateValue === 'string') {
    return resolveTemplateString(templateValue, root);
  }
  const resolver = parsers[templateValue.type];
  return resolver(templateValue, root);
};
