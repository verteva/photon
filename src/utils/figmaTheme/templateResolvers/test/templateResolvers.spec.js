import { replaceTemplateValue } from '../replaceTemplateValue';

import {
  basicRoots,
  compositionRoots,
  colorRoot,
  spacingRoot,
  boxShadowRoot,
  brokenRoot,
} from './themeRoots';

describe('figmaTheme/templateResolvers/', () => {
  // run the same test for all basic parsers.
  Object.keys(basicRoots).forEach(rootName => {
    const keyName = rootName.replace('Root', '');
    const root = basicRoots[rootName];
    describe(`basic parser - ${keyName}`, () => {
      it('parses a terminating value', () => {
        const result = replaceTemplateValue(root[keyName].primary, root);

        expect(result).toEqual(root[keyName].primary.value);
      });

      it('parses a template string', () => {
        const result = replaceTemplateValue(root[keyName].secondary, root);

        expect(result).toEqual(root[keyName].primary.value);
      });
    });
  });

  describe('basic parser - nested values', () => {
    it('parses a nested template string', () => {
      const result = replaceTemplateValue(colorRoot.color.tertiary, colorRoot);

      expect(result).toEqual('#000000 50% #000000 50%');
    });
  });

  describe('basic parser - math in parsers', () => {
    it('parses a nested template with a math equation', () => {
      const result = replaceTemplateValue(
        spacingRoot.spacing.tertiary,
        spacingRoot
      );

      expect(result).toEqual('20');
    });
  });

  describe('color parser - rgba values', () => {
    it('parses a rgba template string', () => {
      const result = replaceTemplateValue(
        colorRoot.color.quaternary,
        colorRoot
      );

      expect(result).toEqual('rgba(0, 0, 0, 0.5)');
    });
  });

  Object.keys(compositionRoots).forEach(rootName => {
    const keyName = rootName.replace('Root', '');
    const root = compositionRoots[rootName];
    describe(`compositionParser - ${keyName}`, () => {
      it('parses a terminating value', () => {
        const result = replaceTemplateValue(root[keyName].primary, root);

        expect(result).toEqual(root[keyName].primary.value);
      });

      it('parses a string template value', () => {
        const result = replaceTemplateValue(root[keyName].secondary, root);
        expect(result).toEqual(root[keyName].primary.value);
      });

      it('parses a nested template value', () => {
        const result = replaceTemplateValue(root[keyName].tertiary, root);

        expect(result).toEqual(root[keyName].primary.value);
      });

      it('parses a nested template with templated strings', () => {
        const result = replaceTemplateValue(root[keyName].danger, root);

        expect(result).toEqual(root[keyName].primary.value);
      });
    });
  });

  describe('boxShadowParser', () => {
    const root = { ...colorRoot, ...boxShadowRoot };
    it('parses a terminating value', () => {
      const result = replaceTemplateValue(root.boxShadow.primary, root);

      expect(result).toEqual('0px 0px 0px 0px #000000');
    });

    it('parses a string template value', () => {
      const result = replaceTemplateValue(root.boxShadow.secondary, root);

      expect(result).toEqual('0px 0px 0px 0px #000000');
    });

    it('parses a nested template value', () => {
      const result = replaceTemplateValue(root.boxShadow.tertiary, root);

      expect(result).toEqual('0px 0px 0px 0px #000000');
    });

    it('parses a nested template with templated strings', () => {
      const result = replaceTemplateValue(root.boxShadow.danger, root);

      expect(result).toEqual('0px 0px 0px 0px #000000');
    });

    it('parses an array box shadow value', () => {
      const result = replaceTemplateValue(root.boxShadow.array, root);

      expect(result).toEqual('0px 0px 0px 0px red, 0px 0px 0px 0px blue');
    });
  });

  describe('errors', () => {
    it('Throws an error when trying to parse objects into strings', () => {
      expect(() =>
        replaceTemplateValue(brokenRoot.other.primary, brokenRoot)
      ).toThrowErrorMatchingSnapshot();
    });

    it('logs an error when trying to parse objects that do not exist', () => {
      jest.spyOn(console, 'error').mockImplementation();

      replaceTemplateValue(brokenRoot.other.secondary, brokenRoot);

      // assert the expected warning
      expect(console.error).toHaveBeenCalledWith(
        expect.stringContaining('value is empty for match')
      );
    });

    it('Throws an error when trying to parse an object without a type key', () => {
      expect(() =>
        replaceTemplateValue(brokenRoot.other.tertiary, brokenRoot)
      ).toThrowErrorMatchingSnapshot();
    });

    it("Throws an error when trying to parse an key that's not a string nor object", () => {
      expect(() =>
        replaceTemplateValue(brokenRoot.other.quaternary, brokenRoot)
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
