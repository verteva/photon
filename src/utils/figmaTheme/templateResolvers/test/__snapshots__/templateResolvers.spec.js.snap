// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`figmaTheme/templateResolvers/ errors throws error when trying to parse an key thats not a string nor object 1`] = `"Invalid template value, must be a string or an object with a type property"`;

exports[`figmaTheme/templateResolvers/ errors throws error when trying to parse an object without a type key 1`] = `"Invalid template value, cant parse objects without a type"`;

exports[`figmaTheme/templateResolvers/ errors throws error when trying to parse objects into strings 1`] = `
"Error: Template string is not valid. You tried to parse object value into a string. 

- templateString: {typography.primary} {color.primary} 
	- match: typography.primary 
	- value: [object Object]"
`;

exports[`figmaTheme/templateResolvers/ errors throws error when trying to parse objects that do not exist 1`] = `"TypeError: resolver is not a function"`;

exports[`figmaTheme/templateResolvers/ errors throws error when trying to parse types that do not exist 1`] = `"resolver is not a function"`;
